iOS platform view mutation XCUITests

cyanglaz opened the pull request:
* Added scenarios for all the available mutations on iOS platform view. Including: clipRect, clipRRect, clipPath, transform, opacity
* Added screenshot tests for each scenarios that added.
* for each test(including the existing test), added code to wait until the platform view is visible on the screen.
* Added golden for simulator iphone se
* ClippingView was previously size (0, 0). The XCUITest ignore the ClippingView because of this. Updated the ClippingView to the same size as flutterView so it can be found by the XCUITest.
* Factored the `platform_view.dart` so it is easier to add a new platform related scenario
* Factored some util methods out of `PlatformViewUITests.m` so it is easier to add new platform view related UITests

@dnfield let me know if you prefer smaller patches.

All the platform view mutation updates will be unblocked after landing this.

dnfield commented:
@jmagman  might have some good tips about structuring the Unit tests and generally reviewing the Objective C :)

cyanglaz commented:
Updated the PR!

@dnfield 
I ended up mostly followed with @jmagman's suggestion on the objc side. 
On the dart side, I updated with your suggestion and also added a little drawing to show what would the clip path look like :). Also added Matrix4 deps for transform. And refactored more code in onBeginFrame into the parent class including adding a picture on top of the platform view

@jmagman 
I improvised a bit and created a `PlatformViewGoldenTestManager` along with the `GoldenImage` class.
`GoldenImage` handles things related to golden, such as getting a UIImage from golden name, construct the golden name, and compare the golden image to a UIImage(screenshot).
`PlatformViewGoldenTestManager` contains a map mapping an id with the run arguments. So it chose what run args to use when initialized with an id. It also creates a `GoldenImage` based on the id., which the id is used as the prefix of the goldenImageName.

dnfield commented on testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewGoldenTestManager.m:
```diff
@@ -0,0 +1,56 @@
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+      [self setLaunchArgsMap:@{
+        @"platform_view" : @[ @"--platform-view" ],
```
Nit - maybe if we defined this in the header for AppDelegate we could reuse it in both places?

cyanglaz replied:
I don't think UITest can include any symbols from the app target. cc @jmagman to confirm. Not sure if there is a way to share code between app target and uitest target.

dnfield commented on testing/scenario_app/lib/src/platform_view.dart:
```diff
@@ -26,21 +27,164 @@ List<int> _to64(num value) {
+    // The left side of the rectangle becomes a symmetric curve towards the left.
+    // The right side of the rectangle becomes a double curve. From top of bottom of the curve, it goes left then right.
+    //     _______
+    //    |      |
```
Nit: Maybe just a reference to the golden image for this, or a link to it on github instead of the ASCII art.

dnfield commented on testing/scenario_app/lib/src/platform_view.dart:
```diff
@@ -26,21 +27,164 @@ List<int> _to64(num value) {
+
+    builder.pushOffset(0, 0);
+    final Matrix4 matrix4 = Matrix4.identity();
+    matrix4.rotateZ(1.0);
```
Nit: 

```
final Matrix4 matrix4 = Matrix4.identity()
  ..rotateZ(1.0)
  ..scale(0.5, 0.5, 1.0)
  ..translate(1000.0, 100.0, 0.0);
```

dnfield commented on testing/scenario_app/lib/src/platform_view.dart:
```diff
@@ -26,21 +27,164 @@ List<int> _to64(num value) {
+    matrix4.scale(0.5, 0.5, 1.0);
+    matrix4.translate(1000.0, 100.0, 0.0);
+
+    final Float64List matrix4_64 =  Float64List.fromList(matrix4.storage);
```
You don't need the copy here - just do `builder.pushTransform(matrix4.storage)`

cyanglaz replied:
matrix4.storage is a Float32List, `builder.pushTransform(matrix4.storage)` throws a compile error.

dnfield replied:
Ahh - make sure you import `vector_math_64.dart` :)

dnfield commented on testing/scenario_app/lib/src/platform_view.dart:
```diff
@@ -26,21 +27,164 @@ List<int> _to64(num value) {
       _valueString,
-      'create'.length, // this is safe as long as these are all single byte characters.
+      'create'
+          .length, // this is safe as long as these are all single byte characters.
```
Please move this back up to line 186 - it's ok if the comment goes over 80. Maybe just ditch the comment, or rephrase as `// this won't work if we use multibyte characters`

dnfield commented on testing/scenario_app/pubspec.yaml:
```diff
@@ -8,3 +8,4 @@ dependencies:
    path: ../../../out/host_debug_unopt/gen/dart-pkg/sky_engine
  sky_services:
    path: ../../../out/host_debug_unopt/gen/dart-pkg/sky_services
+ vector_math: any
```
Please use a specific version for vector_math, e.g. `^2.0.8`

dnfield commented:
Looking better! A few comments on the Dart side, I'll defer to @jmagman to finish out the ObjC review but it looks better overall.

dnfield commented on testing/scenario_app/lib/src/platform_view.dart:
```diff
@@ -6,6 +6,7 @@ import 'dart:convert';
 import 'dart:io';
 import 'dart:typed_data';
 import 'dart:ui';
+import 'package:vector_math/vector_math.dart';
```
you want `import 'package:vector_math/vector_math_64.dart';`

jmagman commented on testing/scenario_app/ios/Scenarios/ScenariosUITests/GoldenImage.m:
```diff
@@ -0,0 +1,101 @@
+  return self;
+}
+
+- (BOOL)compareGoldenToImage:(UIImage*)image {
```
Override `-isEqual`: and `-hash` instead.

jmagman replied:
> Override `-isEqual`: and `-hash` instead.

Oh actually never mind, that would require you to make a GoldenImage out of the screen shot.  This is fine.

jmagman commented on testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewGoldenTestManager.m:
```diff
@@ -0,0 +1,56 @@
+
+static NSDictionary* _launchArgsMap;
+
++ (void)initialize {
```
Overriding `+initialize` is very frowned upon.  You can make the static dictionary a local variable in -initWithIdentifier.  Or create the tested golden out of the launch args.  Or make it an `extern` somewhere to be used by AppDelegate as @dnfield suggests.

jmagman replied:
> Or make it an `extern` somewhere to be used by AppDelegate as @dnfield suggests.

@cyanglaz correctly points out this isn't possible in a XCUITest.

dnfield replied:
Ahh right.

dnfield commented on testing/scenario_app/lib/src/platform_view.dart:
```diff
@@ -26,21 +27,155 @@ List<int> _to64(num value) {
+
+    builder.pushOffset(0, 0);
+    final Matrix4 matrix4 = Matrix4.identity()
+    ..rotateZ(1.0)
```
nit agian: indent these :)

cyanglaz commented:
@jmagman 

Updated the objc code. I think it is ready for another review.

I ended up adding duplicate maps in AppDelegate and PlatformViewGoldenTestManager.

jmagman commented on testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m:
```diff
@@ -19,31 +19,55 @@ - (BOOL)application:(UIApplication*)application
+    };
+  });
+  BOOL hasGoldenLaunchArg = NO;
+  for (NSString* key in launchArgsMap.allKeys) {
```
Use fast-enumeration for iterating dictionaries instead of allocating a new array and then doing a value lookup later.
```
__block BOOL hasGoldenLaunchArg = NO;
[launchArgsMap enumerateKeysAndObjectsUsingBlock:^(NSString* argument, NSString* testName, BOOL *stop) {
   if ([[[NSProcessInfo processInfo] arguments] containsObject:argument]) {
     [self readyContextForPlatformViewTests:testName];
     hasGoldenLaunchArg = YES;
     *stop = YES;
  }
}];
```

jmagman commented on testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m:
```diff
@@ -19,31 +19,55 @@ - (BOOL)application:(UIApplication*)application
-    self.window.rootViewController = flutterViewController;
-  } else {
+  // The launchArgsMap should match the one in the `PlatformVieGoldenTestManager`.
+  static NSDictionary* launchArgsMap;
```
`application:didFinishLaunchingWithOptions:` will only be called once, so you don't need the overhead of the static and dispatch_once.
Also, always use generics.
```
NSDictionary<NSString*, NSString*>* launchArgsMap;
```
